packages/cart/src/services/cart-module.ts

```diff
@@ -658,4 +668,184 @@ export default class CartModuleService implements ICartModuleService {
+    let addedAdjustments: LineItemAdjustmentLine[] = []
+    if (isString(cartIdOrData)) {
+      const cart = await this.retrieve(
+        cartIdOrData,
+        { select: ["id"], relations: ["items"] },
+        sharedContext
+      )
+
+      const lineIds = cart.items?.map((item) => item.id)
+
+      for (const adj of adjustments || []) {
+        if (!lineIds?.includes(adj.item_id)) {
+          throw new MedusaError(
+            MedusaError.Types.INVALID_DATA,
+            `Line item with id ${adj.item_id} does not exist on cart with id ${cartIdOrData}`
+          )
+        }
+      }
+
+      addedAdjustments = await this.lineItemAdjustmentService_.create(
+        adjustments as CartTypes.CreateLineItemAdjustmentDTO[],
+        sharedContext
+      )
+    } else {
+      const data = Array.isArray(cartIdOrData) ? cartIdOrData : [cartIdOrData]
+
+      addedAdjustments = await this.lineItemAdjustmentService_.create(
+        data as CartTypes.CreateLineItemAdjustmentDTO[],
+        sharedContext
+      )
+    }
+
+    return await this.baseRepository_.serialize<
+      CartTypes.LineItemAdjustmentLineDTO[]
+    >(addedAdjustments, {
+      populate: true,
+    })
+  }
+
+  @InjectTransactionManager("baseRepository_")
+  async setLineItemAdjustments(
+    cartId: string,
+    adjustments: (
+      | CartTypes.CreateLineItemAdjustmentDTO
+      | CartTypes.UpdateLineItemAdjustmentDTO
+    )[],
+    @MedusaContext() sharedContext: Context = {}
+  ): Promise<CartTypes.LineItemAdjustmentLineDTO[]> {
+    const cart = await this.retrieve(
+      cartId,
+      { select: ["id"], relations: ["items.adjustments"] },
+      sharedContext
+    )
+
+    const lineIds = cart.items?.map((item) => item.id)
+
+    const existingAdjustments = await this.listLineItemAdjustments(
+      { item_id: lineIds },
+      { select: ["id"] },
+      sharedContext
+    )
+
+    let toUpdate: CartTypes.UpdateLineItemAdjustmentDTO[] = []
+    let toCreate: CartTypes.CreateLineItemAdjustmentDTO[] = []
+    for (const adj of adjustments) {
+      if ("id" in adj) {
+        toUpdate.push(adj as CartTypes.UpdateLineItemAdjustmentDTO)
+      } else {
+        toCreate.push(adj as CartTypes.CreateLineItemAdjustmentDTO)
+      }
+    }
+
+    const adjustmentsSet = new Set(toUpdate.map((a) => a.id))
+
+    const toDelete: CartTypes.LineItemAdjustmentLineDTO[] = []
+
+    // From the existing adjustments, find the ones that are not passed in adjustments
+    existingAdjustments.forEach((adj: CartTypes.LineItemAdjustmentLineDTO) => {
+      if (!adjustmentsSet.has(adj.id)) {
+        toDelete.push(adj)
+      }
+    })
+
+    await this.lineItemAdjustmentService_.delete(
+      toDelete.map((adj) => adj!.id),
+      sharedContext
+    )
+
+    const [result] = await promiseAll([
+      this.lineItemAdjustmentService_.create(toCreate, sharedContext),
+      this.lineItemAdjustmentService_.update(toUpdate, sharedContext),
```

Comment:
question: should we use an upsert to handle both automatically?

Reply:
Will not change this now, as we have the base method coming.